Assert more invariants on ValueRanges

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #95063
* #94985
* #94944
* #94939
* __->__ #94906

The main new invariant is lower/upper must be a Sympy expression of some sort (filtered through `simple_sympify`). There are some simpler sanity checks (mostly making sure the range is well formed). There is a type confusion problem (it's not immediately obvious if a range is for float/int/bool) but we aren't going to solve this for now as it is more complicated.

Billing of changes:

* ValueRanges.wrap() now accepts sympy expressions
* ValueRanges now accepts non-sympy expressions and will sympyify them appropriately. Rewrite calls to ValueRanges to not sympify manually as it is unnecessary
* Don't attempt to test sqrt(-1)
* Add ValuesRanges.unknown() which gives -oo, oo bounds, and rewrite direct calls to -math.inf, math.inf to use it
* Make multiply work between ValueRanges.unknown() and ValueRanges.wrap(0)
* Consistently use sympy.oo instead of math.inf

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
Assert more invariants on ValueRanges

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Assert more invariants on ValueRanges"


Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

eellison commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -9,10 +10,31 @@
+    elif isinstance(e, BooleanAtom):
+        pass
+    else:
+        raise AssertionError(f"not simple sympy type {type(e)}")
```
We will eventually have free symbols when we run this on dynamic shapes to optimize indexing to int32, since we will be creating expressions from the inputs. But this is fine for now.

ezyang replied:
have comments about this now

ezyang commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -9,10 +10,31 @@
+def assert_simple_sympy(e):
+    if isinstance(e, sympy.Expr):
+        assert not e.free_symbols, f"has free variables {ee}"
+        assert e != sympy.nan
```
I think I prefer not converting nan to inf, as it is not always the correct thing to do (for example, 0 * -Inf for a lower bound should go to -Inf not Inf). Better to force the call sites to do it. Maybe if there's a pattern we can factor it later.

ezyang commented:
cc @nunoplopes

ezyang commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -187,7 +209,13 @@ def add(a, b):
     @staticmethod
     def mul(a, b):
-        return ValueRanges.coordinatewise_monotone_map(a, b, operator.mul)
+        def safe_mul(a, b):
```
It's correct for integer analysis, and I don't care about float analysis

ezyang committed:
Update on "Assert more invariants on ValueRanges"


Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

eellison commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -9,21 +10,76 @@
+        if isinstance(lower, sympy.Expr):
+            assert lower <= upper
+        else:
+            assert not (lower is sympy.true and upper is sympy.false)
```
why do we need to assert this ?

ezyang replied:
So this is morally equivalent to lower <= upper, but sympy booleans don't have <= defined on them so I have to do it separately. Probably can make this more clear; would you prefer something like `valid_range(a, b)`

eellison commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -9,21 +10,76 @@
-        # TODO This needs to be generalised if lower/upper are sympy.Expr
-        assert not isinstance(x, sympy.Expr)
+        x = simple_sympify(x)
         return self.lower <= x <= self.upper
```
i think we should still assert that the output is a bool. wydt ? I think user side people are going to not expect that this might return a value which cannot be used as a boolean. 

```
>>> import sympy
>>> from sympy import *
>>> x = Symbol("x")
>>> y = Symbol("y")
>>> bool(x <= y)
TypeError: cannot determine truth value of Relational
```

ezyang commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -9,21 +10,76 @@
+        lower = simple_sympify(lower)
+        upper = simple_sympify(upper)
+        # We don't support point-ranges on floating point inf
+        assert lower != sympy.oo
```
Lower can be -oo; it is only illegal to be oo

ezyang committed:
Update on "Assert more invariants on ValueRanges"


The main new invariant is lower/upper must be a Sympy expression of some sort (filtered through `simple_sympify`). There are some simpler sanity checks (mostly making sure the range is well formed). There is a type confusion problem (it's not immediately obvious if a range is for float/int/bool) but we aren't going to solve this for now as it is more complicated.

Billing of changes:

* ValueRanges.wrap() now accepts sympy expressions
* ValueRanges now accepts non-sympy expressions and will sympyify them appropriately. Rewrite calls to ValueRanges to not sympify manually as it is unnecessary
* Don't attempt to test sqrt(-1)
* Add ValuesRanges.unknown() which gives -oo, oo bounds, and rewrite direct calls to -math.inf, math.inf to use it
* Make multiply work between ValueRanges.unknown() and ValueRanges.wrap(0)
* Consistently use sympy.oo instead of math.inf

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Assert more invariants on ValueRanges"


The main new invariant is lower/upper must be a Sympy expression of some sort (filtered through `simple_sympify`). There are some simpler sanity checks (mostly making sure the range is well formed). There is a type confusion problem (it's not immediately obvious if a range is for float/int/bool) but we aren't going to solve this for now as it is more complicated.

Billing of changes:

* ValueRanges.wrap() now accepts sympy expressions
* ValueRanges now accepts non-sympy expressions and will sympyify them appropriately. Rewrite calls to ValueRanges to not sympify manually as it is unnecessary
* Don't attempt to test sqrt(-1)
* Add ValuesRanges.unknown() which gives -oo, oo bounds, and rewrite direct calls to -math.inf, math.inf to use it
* Make multiply work between ValueRanges.unknown() and ValueRanges.wrap(0)
* Consistently use sympy.oo instead of math.inf

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]
